### DISCUSSION TITLE
Storing unsaved Models in Session with their relations

### DISCUSSION THREAD

**Discussion by gianks** [phalcon/cphalcon#15662] [Q&A]
Hi,
putting a Model in session works as expected even if the model has never been saved.
The id of the model will still be null when the object is retrieved from the session and the object save succeeds as normal across requests.

If a another Model object is created and set on the first one via a relationship (thru a setter or a magic property) and the two are not saved before storing the first object in session, the relationship is lost when the object is retrieved.

I have not investigated why this happens (if it's because storing object in session does not support recursion or because the relationship setter does not store the second object within the other) but this leads to ugly code when data is populated and prepared in a request for the subsequent save in a following request.

What is causing this behavior? Is there a practical workaround to store the full chain of unsaved object using just the one which would be finally saved?

Hope I expressed clearly the problem, seemed easier to explain before i started writing this :)

**Comment by quasipickle** [comment]
Likely the relationship isn't saved because of the reason you suspect - storing in $_SESSION isn't recursive.

I can't think of a reason why you would need to _create_ an object on one page, but _save_ it on another.  You might want to take a look at your design and see if it can be improved.  If not, how about not storing the object itself in $_SESSION, but just its properties.  Then you can re-create it–and any related objects– on the page where the object gets saved.

**Reply by gianks** [reply]
Well, i get your point, sure it's not a common scenario but i can see occasions where a multistep process organizes a bounce of temporary data and finally save all in the DB just at the end.
For example, when there is the need to obtain a confirmation and extra data from an external service before saving multiple related models made by user input, you may want to preserve the latter in case of failure to ease a new attempt.
Clearly the objects (models) could be saved in the DB before moving to the next step but this would lead to a certain amount of unnecessary load on the DB to retrieve each time the records and, when the process gets interrupted, introducing also the need to perform some sort of cleanup of such records and therefore of tracking their validity (based on time, i guess).

I understand you can store the properties but this makes useless the ease provided by Forms, Models and Relations, when combined together: it's essentially the same as working with plain arrays more than objects.

I guess the answer is that Phalcon does not mean to provide a way to support such cases.

**Reply by quasipickle** [reply]
I don't know if this will work, but you could try `serialize()`-ing before storing in `$_SESSION`
